### ISSUE TITLE
Backups/Restore functionality 

### ISSUE THREAD

**Issue by p10tyr** [NginxProxyManager/nginx-proxy-manager#2109]
**Is your feature request related to a problem? Please describe.**
I have just started using nginxmanager quite a lot now with some production services too.
I did the easy setup with no external databases so I am wondering how to export/import?

**Describe the solution you'd like**
Export / Backup 
Import / Restore

**Describe alternatives you've considered**
Don't know.. go into the docker container and hack into the sqllite db??

**Additional context**
Good for distater recovery too if we could just export data into a well known format and have it backup up away from the backup

**Comment by Horuyuki** [comment]
Usually you can do this with copying the data and letsencrypt directory of the compose folder, aka the whole npm docker compose folder.

**Comment by sanderlv** [comment]
I would LOVE to see this natively from within the user interface ---> hence that's the reason at all I like this product

**Comment by danielschenk** [comment]
I'm wondering: is atomicity always guaranteed on the sqlite file? In other words: can we be sure that the database on disk is never in an inconsistent state due to ongoing transactions?

If above can't be guaranteed, we need to always stop proxy-manager for a while to be able to pull a safe copy, which can be undesired. In that case, a function to have proxy-manager export a copy (on some kind of trigger which can be automated) would be very helpful. Home Assistant for example has a builtin service for this which can be hooked up to any automation.

**Comment by danielschenk** [comment]
I'm also not familiar with the codebase at all. I just found your issue because I'm missing this kind of function myself too.

As can be read [here](https://www.sqlite.org/atomiccommit.html#section_3_2) atomicity is taken care of, but SQlite uses file locks from the OS for that. If we just read the SQLite file on our own (while proxy manager is also running) we can't be sure that blocks of the file are actually being written because we bypass SQLite's locking protocol. I've noticed that even without changing settings on my instance, the database file gets frequent updates, so I think it's too big of a risk to just read it while proxy manager is running.

Hence I think your feature request is very useful. Why a CSV and not just a copy of the SQLite file?

Regarding the certificates, I think just re-requesting them and not including them in a backup would indeed be the easiest way to solve this.

**Comment by TicTac9k1** [comment]
> Usually you can do this with copying the data and letsencrypt directory of the compose folder, aka the whole npm docker compose folder.

I would, but copying or tarring gives me permissions errors I don't know for sure are relevant for a successful export.

https://i.imgur.com/tKfJ3BV.png

Using the jc21/nginx-proxy-manager:latest docker image

**Comment by smaccona** [comment]
@danielschenk you are absolutely correct that copying the SQLite database file while the database is online is not safe from corruption. SQLite does have several built-in ways to backup a running database:
- The `clone` command, but this is slow and locks the entire database for writes while it's running (you can still read).
- The `dump` command, which dumps the schema and data as SQL to a text file. This is typically faster than `clone`, but still not "fast", and it also locks the database for writes.
- You can use `vacuum into` to defragment the database into another file. This is faster than the first two, but still not very fast and also locks the database for writes.
- The `backup` command, which is faster than all the others, and doesn't lock the whole database (it does perform locks, but it releases them and yields to other connections).
- There's also an API for performing backups (see https://sqlite.org/c3ref/backup_finish.html), but I don't know if it's supported in the `sqlite3` package being used here (but you might be able to simply call the `backup` command by using `exec()` and passing `.backup`).

**Comment by sanderlv** [comment]
Dream on...

**Comment by sanderlv** [comment]
Comment to keep it open

**Comment by sanderlv** [comment]
Stale prevent

**Comment by durzo** [comment]
I've implemented backup and restore via #5183 and would love for people to test it out and provide feedback on the PR.

**Comment by sanderlv** [comment]
> I've implemented backup and restore via [#5183](https://github.com/NginxProxyManager/nginx-proxy-manager/pull/5183) and would love for people to test it out and provide feedback on the PR.

How?

**Comment by durzo** [comment]
> How?

Well, you can't, because the devs aren't open to contributions, only adding lang strings and keeping the project limping along instead of fixing bugs present for months and adding features. no wonder npm-plus exists.

I'm closing the pull request to prevent future confusion.

**Comment by sanderlv** [comment]
Hmz... thats a pity... it npm plus equal to this npm? And easy to migrate?

**Comment by JohnWallerUK** [comment]
Why don't we have a save and restore????

**Comment by samwathegreat** [comment]
We really need a backup / restore function.  Really bad.

**Comment by sanderlv** [comment]
Comment
